File: templates/Virtual-Assistant-Template/typescript/generator-botbuilder-assistant/generators/skill/templates/customSkill/src/adapters/_customSkillAdapter.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { TelemetryClient } from 'applicationinsights';
import {
    AutoSaveStateMiddleware,
    BotFrameworkAdapter,
    BotFrameworkAdapterSettings,
    ConversationState,
    ShowTypingMiddleware,
    StatePropertyAccessor,
    UserState } from 'botbuilder';
import { DialogState } from 'botbuilder-dialogs';
import {
    SkillContext,
    SkillHttpAdapter,
    SkillHttpBotAdapter,
    SkillMiddleware } from 'botbuilder-skills';
import {
    EventDebuggerMiddleware,
    SetLocaleMiddleware,
    TelemetryLoggerMiddleware} from 'botbuilder-solutions';
import { IBotSettings } from '../services/botSettings';

export class CustomSkillAdapter extends SkillHttpBotAdapter {

    constructor(
        settings: Partial<IBotSettings>,
        userState: UserState,
        conversationState: ConversationState,
        telemetryClient: TelemetryClient,
        skillContextAccessor: StatePropertyAccessor<SkillContext>,
        dialogStateAccesor: StatePropertyAccessor<DialogState>
    ) {
        super(
            telemetryClient
        );

        this.use(new TelemetryLoggerMiddleware(telemetryClient, true));
        // Currently not working https://github.com/Microsoft/botbuilder-js/issues/853#issuecomment-481416004
        // this.use(new TranscriptLoggerMiddleware(this.transcriptStore));
        // Typing Middleware (automatically shows typing when the bot is responding/working)
        this.use(new ShowTypingMiddleware());
        let defaultLocale: string = 'en-us';
        if (settings.defaultLocale !== undefined) {
            defaultLocale = settings.defaultLocale;
        }
        this.use(new SetLocaleMiddleware(defaultLocale));
        this.use(new EventDebuggerMiddleware());
        // Use the AutoSaveStateMiddleware middleware to automatically read and write conversation and user state.
        this.use(new AutoSaveStateMiddleware(conversationState, userState));
        this.use(new SkillMiddleware(conversationState, skillContextAccessor, dialogStateAccesor));
    }
}

File: skills/csharp/experimental/weatherskill/Adapters/CustomSkillAdapter.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Globalization;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Azure;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Skills;
using Microsoft.Bot.Builder.Solutions.Middleware;
using Microsoft.Bot.Builder.Solutions.Responses;
using Microsoft.Bot.Schema;
using WeatherSkill.Responses.Shared;
using WeatherSkill.Services;

namespace WeatherSkill.Adapters
{
    public class CustomSkillAdapter : SkillWebSocketBotAdapter
    {
        public CustomSkillAdapter(
            BotSettings settings,
            UserState userState,
            ConversationState conversationState,
            ResponseManager responseManager,
            IBotTelemetryClient telemetryClient)
            : base(null, telemetryClient)
        {
            OnTurnError = async (context, exception) =>
            {
                CultureInfo.CurrentUICulture = new CultureInfo(context.Activity.Locale);
                await context.SendActivityAsync(responseManager.GetResponse(SharedResponses.ErrorMessage));
                await context.SendActivityAsync(new Activity(type: ActivityTypes.Trace, text: $"Skill Error: {exception.Message} | {exception.StackTrace}"));
                telemetryClient.TrackException(exception);
            };

            Use(new TranscriptLoggerMiddleware(new AzureBlobTranscriptStore(settings.BlobStorage.ConnectionString, settings.BlobStorage.Container)));
            Use(new TelemetryLoggerMiddleware(telemetryClient, logPersonalInformation: true));
            Use(new ShowTypingMiddleware());
            Use(new SetLocaleMiddleware(settings.DefaultLocale ?? "en-us"));
            Use(new EventDebuggerMiddleware());
            Use(new SkillMiddleware(userState, conversationState, conversationState.CreateProperty<DialogState>(nameof(DialogState))));
        }
    }
}